### ISSUE TITLE
Use GroupedData for SmvCDS interface (1.3)

### ISSUE THREAD

**Issue by ninjapapa** [TresAmigosSD/SMV#123]
In SMV 1.1, the core CDS interface is
```scala
srdd.smvApplyCDS(keys: _*)(cds).groupBy(keyColsExpr: _*)(aggrExpr: _*)
```

Basically a CDS defined a data manipulation (defined by ```cds```) within a Data Group which grouped by the first parameter set ```keys```,  then an aggregation is performed on the combine of the original ```keys``` and some additional keys defined by the ```cds``` (```keyColsExpr``` is the combination).

In Spark 1.3 the ```groupBy``` method of ```DataFrame``` (SchemaRDD) was broken to 2 pieces:  ```groupBy``` and ```agg```, where ```groupBy``` returns a ```GroupedData``` object.

With this separation, we can make the CDS interface much more clear. One approach could be
```scala
df.groupBy(keys).smvApplyCDS(cds).agg(aggrExprs)
```

So basically ```smvApplyCDS``` now is a method of ```GroupedData``` and return a ```GroupedData```, where the input ```GroupedData``` is grouped by ```keys``` and the second grouped by ```keys``` plus ```cds``` output keys. That is exactly what a CDS should be doing!

**Comment by ninjapapa** [comment]
Move 1 step further, CDS itself could be defined as a method of ```GroupedData``` and return ```GroupedData```. 

It can be chained together. The client code could look like
```scala
df.groupBy('k1, 'k2).predefinedCds1(...).predefinedCds2(...).smvApplyCds(cds3).agg(...)
```

**Comment by ninjapapa** [comment]
As defined in issue #106, our ultimate CDS interface should look like
```scala
srdd.smvGroupBy('k)(
    Sum('v1) from TimeInLastN('t, 3) as 'nv1,
    Count('v2) from TimeInLastN('t, 6) as 'nv2
)
```

In 1.3, it could be
```scala
df.groupBy('k).select(
  sum('v1) from TimeInLastN('t, 3) as 'nv1,
  count('v2) from TimeInLastN('t, 6) as 'nv2
)
```

It's actually quite possible to implement it like the following

* We don't need ```smvGroupBy``` any more, groupBy in 1.3 is good enough
* ```select``` is a method on ```GroupedData```. Spark has no ```select``` method on ```GroupedData```
* ```from``` is a method on ```Column``` which take ```SmvCDS``` as parameter, and return a ```Column```

**Comment by ninjapapa** [comment]
Maybe our current CDS design overload that concept too much. Let's separate CDS from the GroupedData methods. 

## CDS as a "local" DF 
Let's defined CDS as a DF within each group of a GroupedData. Therefor 
```scala
sum('a) from cds1(...)
```
is equivalent to 
```
cds1(...).agg(sum('a))
```
within a group of a GroupedData.

Put it into the GroupedData context
```scala
df.groupBy('k).select(sum('a) from cds1(...))
```
is equivalent to
```scala
df.groupBy('k).smvApplyCDS(cds1(...)).agg(sum('a))
```

With our original use case in mind, CDS here extends **each single record** in the original GroupedData to a "local" DF, and the ```agg``` get it back to a set of regular variables of the original **single record**. 

## Methods on GroupedData could work without CDS
Consider the use case that

"For each group of the same 'k, get the top 3 records of value 'v"

We can define a GroupedData method, and the client code will look like
```scala
df.groupBy('k).smvTopNRec(3, 'v.desc)
```
Here ```smvTopNRec``` does not need to implemented through CDS, it's imply a method of GroupedData and return a DF.

## Combine CDS with GroupedData methods
Consider a use case that our data is phone call record data (transactions), for each phone number, we want to calculated at every month, how many calls he made on the previous 3 months. 

The tricky part of this problem is that if a phone number made 0 calls in March 2015, there will have no records in 201503 from the original data, also since there could be multiple records in 201502, while our final output should be 1 record per phone per month. 

Because of that we can't just apply a CDS on the original transaction data. 

Let's say the transaction data has 2 columns: ```'key``` and ```'calltime```, we can introduce a GroupedData method and combine with CDS to solve the problem
```scala
df.groupBy('key).
  smvAddAnchor('month, MonthRange(201201, 201403)).
  smvApplyCDS(TimeInLastN(3, 'month, 'calltime)).
  agg(count(1))
```
or 
```scala
df.groupBy('key).
  smvAddAnchor('month, MonthRange(201201, 201403)).
  select(count(1) from TimeInLastN(3, 'month, 'calltime))
```

**Comment by AliTajeldin** [comment]
We need a whiteboard session on this.

**Comment by ninjapapa** [comment]
Implemented new design of CDS.
